### PR TITLE
Tree sitter parsers

### PR DESCRIPTION
**Pull request by msva**
CC: @00-matt @sarnex

New grammars.

Most of them also required in runtime by app-edittors/neovim-9999 (and it seems will be required by next non-patch release too)

Ref: https://github.com/neovim/neovim/blob/master/cmake.deps/deps.txt#L46 and below

### CODE REVIEW

**Comment by MatthewGentoo** [comment]
Thanks!

Think that these new grammars need `test` masking for rustless profiles: https://github.com/gentoo/gentoo/blob/master/profiles/features/wd40/package.use.mask#L79

**Comment by msva** [comment]
@MatthewGentoo done

**Comment by msva** [comment]
Please, leave a note, if all looks good for you, and I'll push it in gentoo myself

(this PR is for review purposes only, I have push access myself, if any)

**Comment by msva** [comment]
> but some of the tests in `tree-sitter-markdown-inline` are failing.
Unfortunately, I have no idea what can I do with that :shrug: 
Can you give some advice?
Maybe just drop markdown-inline grammar?

**Comment by thesamesam** [comment]
First, you check if you can reproduce it.

If you can, you poke at it, and report it upstream?

Did you not run the tests for it before opening the PR?

File: dev-libs/tree-sitter-lua/tree-sitter-lua-0.0.19.ebuild [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+inherit tree-sitter-grammar
+
+DESCRIPTION="Lua grammar for Tree-sitter"
+HOMEPAGE="https://github.com/tree-sitter-grammars/tree-sitter-lua"
+
+SRC_URI="https://github.com/tree-sitter-grammars/${PN}/archive/${TS_PV:-v${PV}}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~ppc ~ppc64"
```
[the remark this answers is not in the document]
Comment (msva) [reply dev-libs/tree-sitter-lua/tree-sitter-lua-0.0.19.ebuild:15]: I did tested that it compiles fine on all of that arches.
Although, I only tested compilation itself, and didn't tested them in co-operation with "clients".
And also I tested ppc and ppc64 on qemu targets (arm and arm64 on "real" boards (`orange pi zero` and `pine a64+`)).
Isn't it enough?
If so - okay, I will remove anything except `~amd64` :shrug:

[the remark this answers is not in the document]
Comment (thesamesam) [reply dev-libs/tree-sitter-lua/tree-sitter-lua-0.0.19.ebuild:15]: Can I ask where you got the impression this was the done thing? This is covered in the quizzes and in the devmanual quite clearly?

**Comment by thesamesam** [comment]
>  I did. But this grammar passes many other tests, and can still be useful.

...? Did you hit the failure, then just not mention it at all? That's bizarre?

**Comment by msva** [comment]
> Did you hit the failure, then just not mention it at all

Well, it is really my faultt, that I forgot to mention that (altough, that is the reason I created PR for: to other co-maintainers re-checked the things I forgot about).


As "bas excuse" (that actually not excusinng, but gives a light on reasons why did itt happened) for that, I can mention that I'm working on this PR for about a month already (by minor steps when getting some spare hours and not-too-tired after work), so at current point I just forgot about this "speciality" of this not-so-necessary package (that was added mostly because it is co-packaged with not-inline version), and remembered that fact only when @MatthewGentoo mentioned that.
And when I hit that with personal testing, IIRC (that was some time ago), I was guided by the logic like (I mean, "my thoughts was"):
```
1. it still does most of its work,
2. the failed tests looks like not so related to things where it will be needed as "inline" (and looks like inherited from non-inline ("main") grammar), # altthough, I'm not a big specialist here :shrug: 
3. this package is far from "important ones", and such fails of tests most probably is not a reason to not package it at all (especially in case I've seen many times packages that tfails tests in the gentoo repo), so I'll mention that later # (here is where I mostly failed)
```

So, yes, it is definittelly my fault in forgetting about it, but I don't agree with "bizzare".
It is just "artifact" of a process how I worked on this, which was hopefuly catched by a process specifically designed for catching things like this :shrug: 

Also, as I said, this package (unlike others in the PR) is not a dependency of neovim, so it is not strictly needed.

Well, I already dropped it in local copy of branch, and will force-push commits without it :shrug: 

(I anyway planned to force-push as there is another change: `neovim` had moved `vim` grammar to `tree-sitter-grammars` organization while I worked on other things)

**Comment by msva** [comment]
@MatthewGentoo I've re-pushed, can you, please, re-review?

Also
@thesamesam @arthurzam @blueness @Flowdalic @leio @NeddySeagoon 
can you please correctly arch-test it on ARM64, please? (I have only tested that it compiles)
I'd very like to push it with at least `~amd64 ~arm64`

**Comment by msva** [comment]
(re-pushed again because of missed "Bug" header in previous force-push)

**Comment by msva** [comment]
by the way, tree-sitter-markdown said to "try passing ALL_EXTENSIONS=1".
So, I tried re-generating the things by calling `ALL_EXTENSIONS=1 tree-sitter generate --no-bindings` in `tree-sitter-inline` directory, and re-run the test, and it passes all the things.
But as it requires calling of `tree-sitter-cli` which is unavailable on rustless profiles, I'm not sure about the correct way to go here.

So, I think, I'll leave the inline grammar and will not add an ebuild for it for now.
In case if it will be needed in future - it can be made from non-inline ebuild with minor modificatitons.

**Comment by thesamesam** [comment]
No, please follow the policy I cited at https://github.com/gentoo/gentoo/pull/35461#discussion_r1497251400. This package does not get special treatment. You get it merged, then you request (re)keywording via a bug.

It is enormous hassle to deviate from that process, _especially_ to handle keywording for things not even in-tree yet.

**Comment by thesamesam** [comment]
> > Did you hit the failure, then just not mention it at all
> 
> Well, it is really my fault, that I forgot to mention that (altough, that is the reason I created PR for: to other co-maintainers re-checked the things I forgot about).
> 

> It is just "artifact" of a process how I worked on this, which was hopefuly catched by a process specifically designed for catching things like this 🤷

In future, you should take care to run the test suites (again before pushing) and to mention any test failures immediately in a comment in the ebuild or similar for further investigation.

We can't do much about people forgetting things - that's why it's good to not let some stuff run on for months indeed.

> Well, I already dropped it in local copy of branch, and will force-push commits without it 🤷

I am not sure why you're using 🤷 so much. It might appear dismissive.

**Comment by msva** [comment]
Btw, @thesamesam, can you, please re-review, and approve the merge if all is okay?

Github shows thaе you have requested changes, but says that can't find corresponding things because of force-push

**Comment by msva** [comment]
By the way, I looked on one of my arm-boards (the one with ubuntu), and found that neovim is packaged with `markdown-inline` there.

```
/usr/lib/arm-linux-gnueabihf/nvim/parser/markdown_inline.so
```

So... Then there is again appears a question: should we package it or not.

Without, for example, Lua grammar, `nvim` fires an error when opening any lua file even with "clean" profile.

Without Markdown grammars (both) `nvim` doesn't fire errors on opening markdown files (at least, for now).  
So, it seems kinda optional for now (even despite upstream bundles botth inlinne and non-inline markdown grammars (among others) on other OSes).

And, if we decide to package (again) inline grammar, would it be okay to regenerate it witth tree-sitter-cli (as upstream suggests) before building? It will definitelly make it unavailable (or failing tests) on ~~wd40~~ rustless architectures...


@MatthewGentoo @sarnex WDYT?

**Comment by MatthewGentoo** [comment]
> And, if we decide to package (again) inline grammar, would it be okay to regenerate it with tree-sitter-cli (as upstream suggests) before building? It will definitelly make it unavailable (or failing tests) on wd40 rustless architectures...

* We could regenerate the parser ourselves with extensions enabled, and upload to dev.gentoo.org. Maybe this is a bad idea though, as it looks like NeoVim doesn't build with the extensions either: https://github.com/neovim/neovim/blob/5b312cd5f67646005312d2ebb2ef0d1bc584dcff/cmake.deps/cmake/MarkdownParserCMakeLists.txt#L17
* We could set RESTRICT=test
* We could patch to delete the failing tests that require extensions

I think the main thing is you are happy it works with your editor :)

Other grammars look good.

**Comment by mattst88** [comment]
Looks like we have newer version of all of these packages in ::gentoo now.
